src/main/java/org/intellij/sdk/codesync/clients/CodeSyncWebSocketClient.java

```diff
@@ -93,23 +105,79 @@ public void sendDiff(DiffFile diffFile, Integer fileId, DataTransmissionHandler
         this.webSocketClientEndpoint.setMessageHandler(message -> {
             if (message.isEmpty()) {
                 System.out.println("Got empty response while authenticating diff.");
-                dataTransmissionHandler.dataTransferStatusCallback(false);
+                dataTransmissionHandler.dataTransferStatusCallback(false, null);
             }
 
             JSONObject response;
             try {
                 response = (JSONObject) JSONValue.parseWithException(message);
                 Long statusCode = (Long) response.get("status");
+                String diffFilePath = (String) response.get("diff_file_path");
                 if (statusCode != 200) {
                     System.out.printf("Diff auth Failed with error: %s.", response.get("error"));
                 }
-                dataTransmissionHandler.dataTransferStatusCallback(statusCode == 200);
+                dataTransmissionHandler.dataTransferStatusCallback(statusCode == 200, diffFilePath);
+            } catch (org.json.simple.parser.ParseException error) {
+                System.out.println("Invalid response from the server.");
+                dataTransmissionHandler.dataTransferStatusCallback(false, null);
+            } catch (ClassCastException error) {
+                System.out.println("Invalid status code.");
+                dataTransmissionHandler.dataTransferStatusCallback(false, null);
+            }
+        });
+
+        this.webSocketClientEndpoint.sendMessage(payload.toJSONString());
+    }
+
+    public void sendDiffs(ArrayList<Pair<Integer, DiffFile>> diffsToSend, DataTransmissionHandler dataTransmissionHandler) throws WebSocketConnectionError {
+        if (!this.isConnected) {
+            throw new WebSocketConnectionError(
+                String.format("Failed to connect to the websocket endpoint at '%s'.}", this.uri.toString())
+            );
+        }
+        JSONArray diffs = new JSONArray();
+
+        for (Pair<Integer, DiffFile> diffFileEntry : diffsToSend) {
+            Integer fileId = diffFileEntry.getFirst();
+            DiffFile diffFile = diffFileEntry.getSecond();
+
+            JSONObject diff = new JSONObject();
+            diff.put("file_id", fileId);
+            diff.put("diff", diffFile.diff);
+            diff.put("is_deleted", diffFile.isDeleted);
+            diff.put("is_rename", diffFile.isRename);
+            diff.put("is_binary", diffFile.isBinary);
+            diff.put("created_at", Utils.formatDate(diffFile.createdAt));
+            diff.put("path", diffFile.fileRelativePath);
+            diff.put("diff_file_path", diffFile.originalDiffFile.getPath());
+
+            diffs.add(diff);
+        }
+
+        JSONObject payload = new JSONObject();
+        payload.put("diffs", diffs);
```

Comment:
Are we sending chunk of diffs in one send call? (i.e. more than 1 diff per sendDiff request?)

Reply:
yes

Reply:
Ok, in JS & brew daemons we are sending 1 diff per send request to handle it independently. However server design is flexible, can accept multiple diffs. But we need to see there is no duplication. e.g. 
1- How many diffs are sent in 1 request?
2- Say if 8/10 are successfuly processed by server but 2/10 encountered an error, what will happen with those 2 diffs?

Reply:
Ah, I see
 ```String diffFilePath = (String) response.get("diff_file_path");``` which means failed diffs will remain in buffer and will be tried again in next iteration. Did I get this right?

Reply:
If this is working well, we probably can use same optimization in JS sides as well where callback is async.

Reply:
Yes, you are correct. I did not see any errors in my testing so I think we are good to implement it in the JS side as well.